_posts/2022-01-18-four-favorite-firmware-debug-tools.md

```diff
@@ -0,0 +1,128 @@
+
+## Segger J-Link
+
+What good is a microcontroller if you can't program it? Segger J-Link debug probes have slowly but surely established themselves as high quality programmers for firmware developers the world over. 
+
+### Nothin' but a J Thang
+
+One of the lovely things about J-Links is the multitude of different software interfaces for their probes. Here's a small sampling of the ways you can chat with your J-Link:
+
+* Command line (J-Link Commander) for daily driver usage
+* GDB server interface
+* Network interfaces, to allow for JTAG access behind an Ethernet switch
+* Factory programmer for speed, verification
+* Raw API to write your own, if one of these 
+
+One of the areas in which Segger really shines is their documentation. Segger distinguishes themselves with high quality user manuals and application notes. All J-Links come with a 400 page user manual that gives detailed explanations of all functionality in associated software. Much as I love free software, it’s a lovely reminder of how paid software incentivizes you to do a really good job on the hard, not-fun parts of the work, like writing good documentation. 
+
+That's not to say that J-Link doesn't play nicely with free software! They are more than capable of interfacing with OpenOCD. Sue Mohielden from River Loop Security put together [this handy guide](https://www.riverloopsecurity.com/blog/2021/07/hw-101-jtag-part3/) showing how to connect to an arbitrary device using a J-Link:
+
+```shell
+sudo ./openocd -s ../tcl -f interface/jlink.cfg -f target/ath79.cfg -c "adapter speed 15000"
+```
+
+All you need to do is sub out the `target/ath79.cfg` argument for whatever target process you are trying to communicate with. You may need to spend some time rummaging around in the `openocd/target` install directory to find the correct file. The exact location of this directory varies from system to system. Mine is available at the following path on my Linux machine: `/usr/share/openocd/scripts/target`. 
+
+This initial configuration process can feel a bit like searching for the proper incantation for a particularly arcane spell. I myself have struggled with getting OpenOCD working several times. Not to worry! After the umpteenth time of getting grumpy with setup, I took down some notes on how to do it for the next time. [I pass it on to you, friends, that your brain sweat be saved for funner problems!](https://cushychicken.github.io/embedded-openocd-gdb/)
+
+### Crash Rules Everything Around Me
+Now, back to talking J-Links! The major drawback of the J-Link ecosystem has to be price. The J-Link Base model clocks in at a hefty $450. The Pro Models top out at just under $1200. That is a _lot_ of cheddar for what is, at the end of the day, a programming tool. However, that price pays for a lot of things that most open source debuggers can't provide:
```

Comment:
Nit- maybe use "money" instead of "cheddar", I like the jovial tone but that colloquialism might be a little unusual for international readers

```suggestion
Now, back to talking J-Links! The major drawback of the J-Link ecosystem has to be price. The J-Link Base model clocks in at a hefty $450. The Pro Models top out at just under $1200. That is a _lot_ of money for what is, at the end of the day, a programming tool. However, that price pays for a lot of things that most open source debuggers can't provide:
```

Reply:
Good note. Can we compromise on "cash"?

Reply:
Perfect! Thank you 😊